File: frontend/src/objects/blog/BlogContent.ts
/** 博客正文品牌类型；用于创建、更新和详情响应。 */
export type BlogContent = string & { readonly __brand: 'BlogContent' }

/** 解析 helper 的返回结果；成功携带品牌值，失败携带可展示错误，不抛异常。 */
type ParseResult<T> = { ok: true; value: T } | { ok: false; error: string }

/** 创建博客正文品牌值；调用前必须完成非空和长度校验。 */
function createBlogContent(value: string): BlogContent {
  /** 注意：这里的 as 只在 parseBlogContent 校验通过后施加品牌类型。 */
  return value as BlogContent
}

/** 将博客正文品牌值还原为字符串；用于 API body 和展示。 */
export function blogContentValue(content: BlogContent): string {
  return content
}

/** 解析博客正文；去除首尾空白并限制最大长度。 */
export function parseBlogContent(rawContent: string): ParseResult<BlogContent> {
  const normalized = rawContent.trim()
  if (!normalized) {
    return { ok: false, error: 'Blog content is required.' }
  }
  if (normalized.length > 200000) {
    return { ok: false, error: 'Blog content must be at most 200000 characters.' }
  }
  return { ok: true, value: createBlogContent(normalized) }
}

File: backend/src/main/scala/domains/blog/objects/BlogContent.scala
package domains.blog.objects

import io.circe.{Decoder, Encoder}


/** 博客正文领域值，保存裁剪后的 Markdown/文本内容。 */
final case class BlogContent(value: String)

/** 提供博客正文 JSON codec 和非空/长度校验。 */
object BlogContent:
  given Encoder[BlogContent] = Encoder.encodeString.contramap(_.value)
  given Decoder[BlogContent] = Decoder.decodeString.emap(parse)

  /** 去除首尾空白，要求正文非空且最多 200000 字符。 */
  def parse(raw: String): Either[String, BlogContent] =
    val normalized = raw.trim
    if normalized.isEmpty then Left("Blog content is required.")
    else if normalized.length > 200000 then Left("Blog content must be at most 200000 characters.")
    else Right(BlogContent(normalized))
